<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Pandas---Basics" data-toc-modified-id="Pandas---Basics-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Pandas - Basics</a></span><ul class="toc-item"><li><span><a href="#Importing-the-module" data-toc-modified-id="Importing-the-module-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Importing the module</a></span></li><li><span><a href="#Loading-the-data" data-toc-modified-id="Loading-the-data-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Loading the data</a></span></li><li><span><a href="#Exploring-the-data" data-toc-modified-id="Exploring-the-data-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Exploring the data</a></span></li><li><span><a href="#Indexing-and-selecting-the-data" data-toc-modified-id="Indexing-and-selecting-the-data-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Indexing and selecting the data</a></span></li><li><span><a href="#Computing-summary-statistics" data-toc-modified-id="Computing-summary-statistics-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Computing summary statistics</a></span></li><li><span><a href="#Checking-for-errors" data-toc-modified-id="Checking-for-errors-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>Checking for errors</a></span></li><li><span><a href="#Sample-exercises" data-toc-modified-id="Sample-exercises-1.7"><span class="toc-item-num">1.7&nbsp;&nbsp;</span>Sample exercises</a></span></li></ul></li></ul></div>

# Pandas - Basics

When wrangling data with Python, we always begin by importing Python packages like pandas so we could access its tools (or methods) which are useful for data manipulation.

**pandas** is a fast, powerful, flexible and easy to use open source data analysis and manipulation tool, built on top of the Python programming language.

*Description from https://pandas.pydata.org/ 

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/e/ed/Pandas_logo.svg/1200px-Pandas_logo.svg.png" width=500/>


## Importing the module

In [1]:
# Import library
import pandas as pd

Note that we use **as** to alias the library. We do this so we won't have to call the whole name of the module in our code.

For example, if I want to create an empty dataframe without alias, I have to call pandas.DataFrame. Meanwhile, if I use **import pandas as pd**, I only need to call pd.DataFrame.

## Loading the data

*Data from [Global Findex Database 2017](https://microdata.worldbank.org/index.php/catalog/3324#metadata-identification).*

The most common source of simple data is csv or comma-separated values file. Pandas can read other file types as well such as:
- read_excel
- read_sql
- read_json

**In terms of reading files**, we usually pass the path of the file we want pandas to read. In this case, since micro_world.csv exists inside the same folder as our notebook, we only need to indicate the file name. However, if it's inside a folder like **data**, we need to pass **'data/micro_world.csv'**.

In [2]:
# Read the data
data = pd.read_csv('micro_world.csv', engine='python')

Pandas tries to determine what dtype to set by analyzing the data in each column. A ```DtypeWarning``` is raised when the dataset read has different dtypes in a column from a file. Recall that there are different dtypes.

We have access to numpy dtypes: `float`, `int`, `bool`, `timedelta64[ns]` and `datetime64[ns]`. Note that the numpy date/time dtypes are not time zone aware. Pandas extends this set of dtypes with its own: `datetime64[ns, <tz>]`, `category`, `Int64`, `string`, `boolean`, etc.

Read the complete reference here: [Pandas dtype reference](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dtypes.html).

## Exploring the data 

In [3]:
# Show the data
data

economy economycode                                    regionwb  \
0       Afghanistan         AFG                                  South Asia   
1       Afghanistan         AFG                                  South Asia   
2       Afghanistan         AFG                                  South Asia   
3       Afghanistan         AFG                                  South Asia   
4       Afghanistan         AFG                                  South Asia   
...             ...         ...                                         ...   
154918     Zimbabwe         ZWE  Sub-Saharan Africa (excluding high income)   
154919     Zimbabwe         ZWE  Sub-Saharan Africa (excluding high income)   
154920     Zimbabwe         ZWE  Sub-Saharan Africa (excluding high income)   
154921     Zimbabwe         ZWE  Sub-Saharan Africa (excluding high income)   
154922     Zimbabwe         ZWE  Sub-Saharan Africa (excluding high income)   

         pop_adult  wpid_random       wgt  female   age  educ  inc_q  ...  \
0       19456060.0    139880545  1.936754       1  18.0   2.0      1  ...   
1       19456060.0    111705622  0.332867       2  25.0   1.0      5  ...   
2       19456060.0    128866217  1.338906       2  40.0   1.0      2  ...   
3       19456060.0    138171768  0.275619       2  45.0   1.0      1  ...   
4       19456060.0    125156190  0.629304       1  30.0   1.0      4  ...   
...            ...          ...       ...     ...   ...   ...    ...  ...   
154918   9467437.0    167986916  0.413192       2  56.0   1.0      3  ...   
154919   9467437.0    121992603  0.576707       1  56.0   1.0      4  ...   
154920   9467437.0    137769624  1.481972       1  36.0   2.0      4  ...   
154921   9467437.0    181215643  0.259968       1  35.0   3.0      5  ...   
154922   9467437.0    133939668  0.791151       1  46.0   2.0      1  ...   

        receive_wages  receive_transfers  receive_pension  \
0                   4                  4                4   
1                   4                  4                4   
2                   4                  4                4   
3                   4                  4                4   
4                   4                  4                4   
...               ...                ...              ...   
154918              4                  4                4   
154919              4                  4                4   
154920              4                  4                4   
154921              1                  4                4   
154922              4                  3                4   

        receive_agriculture  pay_utilities  remittances  pay_onlne  \
0                       2.0              2          5.0        0.0   
1                       4.0              4          5.0        0.0   
2                       4.0              2          5.0        0.0   
3                       4.0              4          5.0        0.0   
4                       4.0              4          5.0        0.0   
...                     ...            ...          ...        ...   
154918                  4.0              4          5.0        0.0   
154919                  4.0              4          2.0        0.0   
154920                  4.0              4          1.0        0.0   
154921                  3.0              1          1.0        0.0   
154922                  4.0              4          3.0        0.0   

        pay_onlne_mobintbuy  pay_cash  pay_cash_mobintbuy  
0                       NaN       0.0                 NaN  
1                       NaN       0.0                 NaN  
2                       NaN       0.0                 NaN  
3                       NaN       0.0                 NaN  
4                       NaN       0.0                 NaN  
...                     ...       ...                 ...  
154918                  NaN       0.0                 NaN  
154919                  NaN       0.0                 NaN  
154920                  

To show the data, you may opt to use **print(data)**, **display(data)** or just plain **data**.

In [4]:
# Display shape
data.shape

(154923, 105)

It means that the data has 154,923 rows and 105 columns. That's a lot!

In [5]:
# Display head
data.head()

economy economycode    regionwb   pop_adult  wpid_random       wgt  \
0  Afghanistan         AFG  South Asia  19456060.0    139880545  1.936754   
1  Afghanistan         AFG  South Asia  19456060.0    111705622  0.332867   
2  Afghanistan         AFG  South Asia  19456060.0    128866217  1.338906   
3  Afghanistan         AFG  South Asia  19456060.0    138171768  0.275619   
4  Afghanistan         AFG  South Asia  19456060.0    125156190  0.629304   

   female   age  educ  inc_q  ...  receive_wages  receive_transfers  \
0       1  18.0   2.0      1  ...              4                  4   
1       2  25.0   1.0      5  ...              4                  4   
2       2  40.0   1.0      2  ...              4                  4   
3       2  45.0   1.0      1  ...              4                  4   
4       1  30.0   1.0      4  ...              4                  4   

   receive_pension  receive_agriculture  pay_utilities  remittances  \
0                4                  2.0              2          5.0   
1                4                  4.0              4          5.0   
2                4                  4.0              2          5.0   
3                4                  4.0              4          5.0   
4                4                  4.0              4          5.0   

   pay_onlne  pay_onlne_mobintbuy  pay_cash  pay_cash_mobintbuy  
0        0.0                  NaN       0.0                 NaN  
1        0.0                  NaN       0.0                 NaN  
2        0.0                  NaN       0.0                 NaN  
3        0.0                  NaN       0.0                 NaN  
4        0.0                  NaN       0.0                 NaN  

[5 rows x 105 columns]

By default, .head() displays the first 5 rows of the dataframe. Note that indexing starts at 0. 😩

In [6]:
# Display tail
data.tail()

economy economycode                                    regionwb  \
154918  Zimbabwe         ZWE  Sub-Saharan Africa (excluding high income)   
154919  Zimbabwe         ZWE  Sub-Saharan Africa (excluding high income)   
154920  Zimbabwe         ZWE  Sub-Saharan Africa (excluding high income)   
154921  Zimbabwe         ZWE  Sub-Saharan Africa (excluding high income)   
154922  Zimbabwe         ZWE  Sub-Saharan Africa (excluding high income)   

        pop_adult  wpid_random       wgt  female   age  educ  inc_q  ...  \
154918  9467437.0    167986916  0.413192       2  56.0   1.0      3  ...   
154919  9467437.0    121992603  0.576707       1  56.0   1.0      4  ...   
154920  9467437.0    137769624  1.481972       1  36.0   2.0      4  ...   
154921  9467437.0    181215643  0.259968       1  35.0   3.0      5  ...   
154922  9467437.0    133939668  0.791151       1  46.0   2.0      1  ...   

        receive_wages  receive_transfers  receive_pension  \
154918              4                  4                4   
154919              4                  4                4   
154920              4                  4                4   
154921              1                  4                4   
154922              4                  3                4   

        receive_agriculture  pay_utilities  remittances  pay_onlne  \
154918                  4.0              4          5.0        0.0   
154919                  4.0              4          2.0        0.0   
154920                  4.0              4          1.0        0.0   
154921                  3.0              1          1.0        0.0   
154922                  4.0              4          3.0        0.0   

        pay_onlne_mobintbuy  pay_cash  pay_cash_mobintbuy  
154918                  NaN       0.0                 NaN  
154919                  NaN       0.0                 NaN  
154920                  NaN       0.0                 NaN  
154921                  NaN       0.0                 NaN  
154922                  NaN       0.0                 NaN  

[5 rows x 105 columns]

Meanwhile, .tail() displays the last 5 rows.

In [7]:
# Display top 20 rows
data.head(20)

economy economycode    regionwb   pop_adult  wpid_random       wgt  \
0   Afghanistan         AFG  South Asia  19456060.0    139880545  1.936754   
1   Afghanistan         AFG  South Asia  19456060.0    111705622  0.332867   
2   Afghanistan         AFG  South Asia  19456060.0    128866217  1.338906   
3   Afghanistan         AFG  South Asia  19456060.0    138171768  0.275619   
4   Afghanistan         AFG  South Asia  19456060.0    125156190  0.629304   
5   Afghanistan         AFG  South Asia  19456060.0    209050144  0.773167   
6   Afghanistan         AFG  South Asia  19456060.0    167258084  0.481786   
7   Afghanistan         AFG  South Asia  19456060.0    147754514  0.564638   
8   Afghanistan         AFG  South Asia  19456060.0    171367518  1.510448   
9   Afghanistan         AFG  South Asia  19456060.0    145691120  0.675545   
10  Afghanistan         AFG  South Asia  19456060.0    160492339  0.280141   
11  Afghanistan         AFG  South Asia  19456060.0    197893595  0.516904   
12  Afghanistan         AFG  South Asia  19456060.0    129005012  0.388837   
13  Afghanistan         AFG  South Asia  19456060.0    185526924  1.665766   
14  Afghanistan         AFG  South Asia  19456060.0    147434413  0.941063   
15  Afghanistan         AFG  South Asia  19456060.0    151537276  0.731453   
16  Afghanistan         AFG  South Asia  19456060.0    117012927  0.264683   
17  Afghanistan         AFG  South Asia  19456060.0    154497258  0.450706   
18  Afghanistan         AFG  South Asia  19456060.0    111819382  0.604836   
19  Afghanistan         AFG  South Asia  19456060.0    122025729  1.387100   

    female   age  educ  inc_q  ...  receive_wages  receive_transfers  \
0        1  18.0   2.0      1  ...              4                  4   
1        2  25.0   1.0      5  ...              4                  4   
2        2  40.0   1.0      2  ...              4                  4   
3        2  45.0   1.0      1  ...              4                  4   
4        1  30.0   1.0      4  ...              4                  4   
5        1  53.0   2.0      1  ...              2                  4   
6        1  36.0   2.0      2  ...              4                  2   
7        1  29.0   2.0      4  ...              2                  4   
8        1  21.0   1.0      5  ...              4                  4   
9        1  75.0   1.0      5  ...              4                  4   
10       1  25.0   1.0      5  ...              4                  4   
11       2  24.0   1.0      5  ...              4                  4   
12       2  31.0   1.0      5  ...              4                  4   
13       1  22.0   1.0      5  ...              2                  4   
14       1  25.0   2.0      2  ...              4                  4   
15       2  30.0   1.0      1  ...              4                  4   
16       2  35.0   2.0      5  ...              4                  4   
17       2  30.0   1.0      5  ...              4                  4   
18       2  24.0   1.0      1  ...              4                  4   
19       2  19.0   2.0      2  ...              4                  4   

    receive_pension  receive_agriculture  pay_utilities  remittances  \
0                 4                  2.0              2          5.0   
1                 4                  4.0              4          5.0   
2                 4                  4.0              2          5.0   
3                 4                  4.0              4          5.0   
4                 4                  4.0              4          5.0   
5                 4                  2.0              4          5.0   
6                 4                  2.0              4          5.0   
7                 4                  4.0              2          1.0   
8                 4                  2.0              4          5.0   
9                 4                  2.0              4          5.0   
10                4                  2.0              2         

You may also opt to display the top n rows by doing .head(n) where n is the number of rows you want to display. Convenient, right?

In [8]:
# Display data info
data.iloc[:,0:20].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 154923 entries, 0 to 154922
Data columns (total 20 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   economy      154923 non-null  object 
 1   economycode  154923 non-null  object 
 2   regionwb     153923 non-null  object 
 3   pop_adult    154923 non-null  float64
 4   wpid_random  154923 non-null  int64  
 5   wgt          154923 non-null  float64
 6   female       154923 non-null  int64  
 7   age          154472 non-null  float64
 8   educ         154913 non-null  float64
 9   inc_q        154923 non-null  int64  
 10  emp_in       153923 non-null  float64
 11  fin2         154923 non-null  int64  
 12  fin3         69925 non-null   float64
 13  fin4         66146 non-null   float64
 14  fin5         84150 non-null   float64
 15  fin6         84150 non-null   float64
 16  fin7         154923 non-null  int64  
 17  fin8         30098 non-null   float64
 18  fin9         86537 non-n

The .info() displays the columns, the number of non-null records per column and the corresponding data type.

In [9]:
# Display summary statistics regarding the data
data.iloc[:,0:20].describe()

pop_adult   wpid_random            wgt         female  \
count  1.549230e+05  1.549230e+05  154923.000000  154923.000000   
mean   6.825448e+07  1.556738e+08       1.000000       1.539578   
std    2.120044e+08  4.038336e+07       0.657137       0.498433   
min    3.739912e+05  1.111100e+04       0.163356       1.000000   
25%    4.335264e+06  1.334725e+08       0.492798       1.000000   
50%    9.386639e+06  1.591577e+08       0.824828       2.000000   
75%    3.051912e+07  1.851128e+08       1.320904       2.000000   
max    1.134634e+09  2.111109e+08       4.010008       2.000000   

                 age           educ          inc_q         emp_in  \
count  154472.000000  154913.000000  154923.000000  153923.000000   
mean       41.842043       1.833707       3.184208       0.627203   
std        17.912444       0.717956       1.422735       0.483550   
min        15.000000       1.000000       1.000000       0.000000   
25%        27.000000       1.000000       2.000000       0.000000   
50%        39.000000       2.000000       3.000000       1.000000   
75%        55.000000       2.000000       4.000000       1.000000   
max        99.000000       5.000000       5.000000       1.000000   

                fin2          fin3          fin4          fin5          fin6  \
count  154923.000000  69925.000000  66146.000000  84150.000000  84150.000000   
mean        1.562609      1.063282      1.298915      1.621034      1.558324   
std         0.529573      0.286836      0.468202      0.498542      0.512465   
min         1.000000      1.000000      1.000000      1.000000      1.000000   
25%         1.000000      1.000000      1.000000      1.000000      1.000000   
50%         2.000000      1.000000      1.000000      2.000000      2.000000   
75%         2.000000      1.000000      2.000000      2.000000      2.000000   
max         4.000000      4.000000      4.000000      4.000000      4.000000   

                fin7          fin8          fin9         fin10  
count  154923.000000  30098.000000  86537.000000  86537.000000  
mean        1.824597      1.159346      1.223477      1.206859  
std         0.436026      0.381386      0.449687      0.440323  
min         1.000000      1.000000      1.000000      1.000000  
25%         2.000000      1.000000      1.000000      1.000000  
50%         2.000000      1.000000      1.000000      1.000000  
75%         2.000000      1.000000      1.000000      1.000000  
max         4.000000      4.000000      4.000000      4.000000

If you want a VERY quick stats overview of your data, use .describe()!

In [10]:
# Display data types
data.dtypes

economy                 object
economycode             object
regionwb                object
pop_adult              float64
wpid_random              int64
                        ...   
remittances            float64
pay_onlne              float64
pay_onlne_mobintbuy    float64
pay_cash               float64
pay_cash_mobintbuy     float64
Length: 105, dtype: object

In [11]:
# Display columns
data.columns

Index(['economy', 'economycode', 'regionwb', 'pop_adult', 'wpid_random', 'wgt',
       'female', 'age', 'educ', 'inc_q',
       ...
       'receive_wages', 'receive_transfers', 'receive_pension',
       'receive_agriculture', 'pay_utilities', 'remittances', 'pay_onlne',
       'pay_onlne_mobintbuy', 'pay_cash', 'pay_cash_mobintbuy'],
      dtype='object', length=105)

## Indexing and selecting the data

Usually, we only want to analyze a subset of the dataset. In this case, we select subset of rows and/or subset of columns.

In [12]:
# Filter to only include the economy column
data['economy']

0         Afghanistan
1         Afghanistan
2         Afghanistan
3         Afghanistan
4         Afghanistan
             ...     
154918       Zimbabwe
154919       Zimbabwe
154920       Zimbabwe
154921       Zimbabwe
154922       Zimbabwe
Name: economy, Length: 154923, dtype: object

Note that one column of a DataFrame is called a Series object!

In [13]:
# Filter economy and economycode columns
data[['economy', 'economycode']]

economy economycode
0       Afghanistan         AFG
1       Afghanistan         AFG
2       Afghanistan         AFG
3       Afghanistan         AFG
4       Afghanistan         AFG
...             ...         ...
154918     Zimbabwe         ZWE
154919     Zimbabwe         ZWE
154920     Zimbabwe         ZWE
154921     Zimbabwe         ZWE
154922     Zimbabwe         ZWE

[154923 rows x 2 columns]

In [14]:
# Filter first two columns using iloc
data.iloc[:, :2]

economy economycode
0       Afghanistan         AFG
1       Afghanistan         AFG
2       Afghanistan         AFG
3       Afghanistan         AFG
4       Afghanistan         AFG
...             ...         ...
154918     Zimbabwe         ZWE
154919     Zimbabwe         ZWE
154920     Zimbabwe         ZWE
154921     Zimbabwe         ZWE
154922     Zimbabwe         ZWE

[154923 rows x 2 columns]

**Note how we use the .iloc**. The first part of the bracket indicates the rows, the second part indicates the columns. 

Here's a couple of examples to guide you:

*   data.iloc[:, :5] - fetch ALL rows for the first 5 columns 
*   data.iloc[:2, :] - fetch the first 2 rows for ALL columns
*   data.iloc[1:3, 5:7] - fetch the 1st row until the 2nd row for 5th to 6th columns





In [15]:
# Select all rows for economy and economycode columns
data.loc[:, 'economy':'economycode']

economy economycode
0       Afghanistan         AFG
1       Afghanistan         AFG
2       Afghanistan         AFG
3       Afghanistan         AFG
4       Afghanistan         AFG
...             ...         ...
154918     Zimbabwe         ZWE
154919     Zimbabwe         ZWE
154920     Zimbabwe         ZWE
154921     Zimbabwe         ZWE
154922     Zimbabwe         ZWE

[154923 rows x 2 columns]

Note that we can use .loc in filtering. In .loc, we can use object type filters in filtering our rows and columns. In this case, we were able to filter columns from economy to economycode. This is more readable than using the .iloc!

In [16]:
# Filter the data to display Philippine data
data[
    data['economy']=='Philippines'
]

economy economycode                                     regionwb  \
114292  Philippines         PHL  East Asia & Pacific (excluding high income)   
114293  Philippines         PHL  East Asia & Pacific (excluding high income)   
114294  Philippines         PHL  East Asia & Pacific (excluding high income)   
114295  Philippines         PHL  East Asia & Pacific (excluding high income)   
114296  Philippines         PHL  East Asia & Pacific (excluding high income)   
...             ...         ...                                          ...   
115287  Philippines         PHL  East Asia & Pacific (excluding high income)   
115288  Philippines         PHL  East Asia & Pacific (excluding high income)   
115289  Philippines         PHL  East Asia & Pacific (excluding high income)   
115290  Philippines         PHL  East Asia & Pacific (excluding high income)   
115291  Philippines         PHL  East Asia & Pacific (excluding high income)   

         pop_adult  wpid_random       wgt  female   age  educ  inc_q  ...  \
114292  70293928.0    180129353  0.261434       2  66.0   1.0      1  ...   
114293  70293928.0    171987745  0.424916       2  24.0   3.0      5  ...   
114294  70293928.0    178228576  0.732469       2  88.0   1.0      4  ...   
114295  70293928.0    193725076  0.388121       2  30.0   2.0      5  ...   
114296  70293928.0    194711242  2.333196       2  33.0   3.0      5  ...   
...            ...          ...       ...     ...   ...   ...    ...  ...   
115287  70293928.0    151988960  1.249522       1  27.0   2.0      1  ...   
115288  70293928.0    138300933  0.697020       1  68.0   1.0      1  ...   
115289  70293928.0    134514859  2.614345       2  18.0   2.0      4  ...   
115290  70293928.0    145841428  1.916246       1  20.0   2.0      4  ...   
115291  70293928.0    208872557  0.435442       1  27.0   1.0      1  ...   

        receive_wages  receive_transfers  receive_pension  \
114292              4                  2                2   
114293              1                  4                4   
114294              4                  2                2   
114295              4                  4                4   
114296              1                  1                4   
...               ...                ...              ...   
115287              2                  4                4   
115288              4                  4                4   
115289              2                  4                4   
115290              2                  4                4   
115291              2                  4                4   

        receive_agriculture  pay_utilities  remittances  pay_onlne  \
114292                  4.0              2          3.0        0.0   
114293                  4.0              4          1.0        0.0   
114294                  4.0              2          5.0        0.0   
114295                  4.0              2          5.0        0.0   
114296                  2.0              2          1.0        0.0   
...                     ...            ...          ...        ...   
115287                  2.0              2          5.0        0.0   
115288                  4.0              4          5.0        0.0   
115289                  4.0              4          2.0        0.0   
115290                  4.0              2          2.0        0.0   
115291                  2.0              4          2.0        0.0   

        pay_onlne_mobintbuy  pay_cash  pay_cash_mobintbuy  
114292                  NaN       0.0                 NaN  
114293                  0.0       1.0                 1.0  
114294                  NaN       0.0                 NaN  
114295                  NaN       0.0                 NaN  
114296                  NaN       0.0                 NaN  
...                     ...       ...                 ...  
115287                  NaN       0.0                 NaN  
115288                  NaN       0.0                 NaN  
115289      

To do filtering in pandas, you need to input a condition inside the bracket.

In this case, the condition is **data['economy'] == 'Philippines'**.

In [17]:
# Filter Philippine correspondents with age < 30
data[
    (data['economy'] == 'Philippines') & (data['age'] < 30)
]

economy economycode                                     regionwb  \
114293  Philippines         PHL  East Asia & Pacific (excluding high income)   
114298  Philippines         PHL  East Asia & Pacific (excluding high income)   
114299  Philippines         PHL  East Asia & Pacific (excluding high income)   
114300  Philippines         PHL  East Asia & Pacific (excluding high income)   
114302  Philippines         PHL  East Asia & Pacific (excluding high income)   
...             ...         ...                                          ...   
115283  Philippines         PHL  East Asia & Pacific (excluding high income)   
115287  Philippines         PHL  East Asia & Pacific (excluding high income)   
115289  Philippines         PHL  East Asia & Pacific (excluding high income)   
115290  Philippines         PHL  East Asia & Pacific (excluding high income)   
115291  Philippines         PHL  East Asia & Pacific (excluding high income)   

         pop_adult  wpid_random       wgt  female   age  educ  inc_q  ...  \
114293  70293928.0    171987745  0.424916       2  24.0   3.0      5  ...   
114298  70293928.0    156608530  1.066926       1  18.0   2.0      2  ...   
114299  70293928.0    204779125  0.776157       2  20.0   2.0      2  ...   
114300  70293928.0    117611794  1.702927       2  23.0   3.0      4  ...   
114302  70293928.0    151574464  1.888463       2  22.0   1.0      2  ...   
...            ...          ...       ...     ...   ...   ...    ...  ...   
115283  70293928.0    139970169  1.455621       2  16.0   2.0      5  ...   
115287  70293928.0    151988960  1.249522       1  27.0   2.0      1  ...   
115289  70293928.0    134514859  2.614345       2  18.0   2.0      4  ...   
115290  70293928.0    145841428  1.916246       1  20.0   2.0      4  ...   
115291  70293928.0    208872557  0.435442       1  27.0   1.0      1  ...   

        receive_wages  receive_transfers  receive_pension  \
114293              1                  4                4   
114298              4                  4                4   
114299              4                  4                4   
114300              2                  4                4   
114302              2                  4                4   
...               ...                ...              ...   
115283              4                  4                4   
115287              2                  4                4   
115289              2                  4                4   
115290              2                  4                4   
115291              2                  4                4   

        receive_agriculture  pay_utilities  remittances  pay_onlne  \
114293                  4.0              4          1.0        0.0   
114298                  4.0              4          5.0        0.0   
114299                  4.0              2          5.0        0.0   
114300                  2.0              2          1.0        0.0   
114302                  4.0              2          2.0        0.0   
...                     ...            ...          ...        ...   
115283                  4.0              4          5.0        0.0   
115287                  2.0              2          5.0        0.0   
115289                  4.0              4          2.0        0.0   
115290                  4.0              2          2.0        0.0   
115291                  2.0              4          2.0        0.0   

        pay_onlne_mobintbuy  pay_cash  pay_cash_mobintbuy  
114293                  0.0       1.0                 1.0  
114298                  0.0       1.0                 1.0  
114299                  NaN       0.0                 NaN  
114300                  NaN       0.0                 NaN  
114302                  NaN       0.0                 NaN  
...                     ...       ...                 ...  
115283                  NaN       0.0                 NaN  
115287                  NaN       0.0                 NaN  
115289      

Here, we used two conditions joined by &. The conditions are also enclosed.

In [18]:
# Filter data from United Kingdom OR United States
data[
    (data['economy']=='United Kingdom') | (data['economy']=='United States')
]

economy economycode           regionwb    pop_adult  \
145916  United Kingdom         GBR  High income: OECD   54072744.0   
145917  United Kingdom         GBR  High income: OECD   54072744.0   
145918  United Kingdom         GBR  High income: OECD   54072744.0   
145919  United Kingdom         GBR  High income: OECD   54072744.0   
145920  United Kingdom         GBR  High income: OECD   54072744.0   
...                ...         ...                ...          ...   
147916   United States         USA  High income: OECD  261642064.0   
147917   United States         USA  High income: OECD  261642064.0   
147918   United States         USA  High income: OECD  261642064.0   
147919   United States         USA  High income: OECD  261642064.0   
147920   United States         USA  High income: OECD  261642064.0   

        wpid_random       wgt  female   age  educ  inc_q  ...  receive_wages  \
145916    139777928  0.448622       2  27.0   3.0      5  ...              1   
145917    199432277  0.758024       2  48.0   3.0      5  ...              1   
145918    204345071  0.667732       1  67.0   2.0      3  ...              4   
145919    186428359  0.876002       2  41.0   3.0      5  ...              1   
145920    157578765  0.883794       2  66.0   3.0      5  ...              4   
...             ...       ...     ...   ...   ...    ...  ...            ...   
147916    129345392  1.158703       2  81.0   2.0      3  ...              1   
147917    156646913  1.560946       1  58.0   2.0      1  ...              4   
147918    209150546  0.905305       2  61.0   2.0      4  ...              1   
147919    207981841  0.684884       2  28.0   3.0      5  ...              1   
147920    144791062  1.412633       1  40.0   2.0      4  ...              1   

        receive_transfers  receive_pension  receive_agriculture  \
145916                  3                4                  NaN   
145917                  4                1                  NaN   
145918                  4                1                  NaN   
145919                  4                4                  NaN   
145920                  4                4                  NaN   
...                   ...              ...                  ...   
147916                  1                1                  NaN   
147917                  1                4                  NaN   
147918                  4                4                  NaN   
147919                  4                4                  NaN   
147920                  4                4                  NaN   

        pay_utilities  remittances  pay_onlne  pay_onlne_mobintbuy  pay_cash  \
145916              1          NaN        NaN                  NaN       NaN   
145917              1          NaN        NaN                  NaN       NaN   
145918              1          NaN        NaN                  NaN       NaN   
145919              1          NaN        NaN                  NaN       NaN   
145920              1          NaN        NaN                  NaN       NaN   
...               ...          ...        ...                  ...       ...   
147916              4          NaN        NaN                  NaN       NaN   
147917              4          NaN        NaN                  NaN       NaN   
147918              1          NaN        NaN                  NaN       NaN   
147919              1          NaN        NaN                  NaN       NaN   
147920              2          NaN        NaN                  NaN       NaN   

        pay_cash_mobintbuy  
145916                 NaN  
145917                 NaN  
145918                 NaN  
145919                 NaN  
145920                 NaN  
...                    ...  
147916                 NaN  
147917                 NaN  
147918                 NaN  
147919                 NaN  
147920                 NaN  

[2005 rows x 105 columns]

In [19]:
# Save the filtered dataframe in a variable
philippine_data = data[
    data['economy'] == 'Philippines'
]

In [20]:
# Show the saved data
philippine_data

economy economycode                                     regionwb  \
114292  Philippines         PHL  East Asia & Pacific (excluding high income)   
114293  Philippines         PHL  East Asia & Pacific (excluding high income)   
114294  Philippines         PHL  East Asia & Pacific (excluding high income)   
114295  Philippines         PHL  East Asia & Pacific (excluding high income)   
114296  Philippines         PHL  East Asia & Pacific (excluding high income)   
...             ...         ...                                          ...   
115287  Philippines         PHL  East Asia & Pacific (excluding high income)   
115288  Philippines         PHL  East Asia & Pacific (excluding high income)   
115289  Philippines         PHL  East Asia & Pacific (excluding high income)   
115290  Philippines         PHL  East Asia & Pacific (excluding high income)   
115291  Philippines         PHL  East Asia & Pacific (excluding high income)   

         pop_adult  wpid_random       wgt  female   age  educ  inc_q  ...  \
114292  70293928.0    180129353  0.261434       2  66.0   1.0      1  ...   
114293  70293928.0    171987745  0.424916       2  24.0   3.0      5  ...   
114294  70293928.0    178228576  0.732469       2  88.0   1.0      4  ...   
114295  70293928.0    193725076  0.388121       2  30.0   2.0      5  ...   
114296  70293928.0    194711242  2.333196       2  33.0   3.0      5  ...   
...            ...          ...       ...     ...   ...   ...    ...  ...   
115287  70293928.0    151988960  1.249522       1  27.0   2.0      1  ...   
115288  70293928.0    138300933  0.697020       1  68.0   1.0      1  ...   
115289  70293928.0    134514859  2.614345       2  18.0   2.0      4  ...   
115290  70293928.0    145841428  1.916246       1  20.0   2.0      4  ...   
115291  70293928.0    208872557  0.435442       1  27.0   1.0      1  ...   

        receive_wages  receive_transfers  receive_pension  \
114292              4                  2                2   
114293              1                  4                4   
114294              4                  2                2   
114295              4                  4                4   
114296              1                  1                4   
...               ...                ...              ...   
115287              2                  4                4   
115288              4                  4                4   
115289              2                  4                4   
115290              2                  4                4   
115291              2                  4                4   

        receive_agriculture  pay_utilities  remittances  pay_onlne  \
114292                  4.0              2          3.0        0.0   
114293                  4.0              4          1.0        0.0   
114294                  4.0              2          5.0        0.0   
114295                  4.0              2          5.0        0.0   
114296                  2.0              2          1.0        0.0   
...                     ...            ...          ...        ...   
115287                  2.0              2          5.0        0.0   
115288                  4.0              4          5.0        0.0   
115289                  4.0              4          2.0        0.0   
115290                  4.0              2          2.0        0.0   
115291                  2.0              4          2.0        0.0   

        pay_onlne_mobintbuy  pay_cash  pay_cash_mobintbuy  
114292                  NaN       0.0                 NaN  
114293                  0.0       1.0                 1.0  
114294                  NaN       0.0                 NaN  
114295                  NaN       0.0                 NaN  
114296                  NaN       0.0                 NaN  
...                     ...       ...                 ...  
115287                  NaN       0.0                 NaN  
115288                  NaN       0.0                 NaN  
115289      

## Computing summary statistics

In pandas, we can compute summary statistics in our sliced dataframe.

In [21]:
# Compute count of correspondents for those in the Philippines
len(philippine_data)

1000

In [22]:
# Compute average age of correspondents in the Philippines
philippine_data['age'].mean()

40.474

In [23]:
# Compute median age of correspondents in the Philippines
philippine_data['age'].median()

37.0

In [24]:
# Compute min age of correspondents in the Philippines
philippine_data['age'].min()

15.0

In [25]:
# Compute min age of correspondents in the Philippines
philippine_data['age'].max()

95.0

In [26]:
# Compute number of respondents per income group
philippine_data['inc_q'].value_counts()

5    228
4    209
3    203
1    192
2    168
Name: inc_q, dtype: int64

According to the **data description**, **inc_q** is the **within-economy household income quintile** (a version of percentile wherein we divide 100% into 5 groups).

Here's the details: 1 - poorest 20%, 2 - second 20%, 3 - middle 20%, 4 - fourth 20%
and 5 - richest 20%.

Based on the data, we have highest number of respondents in the richest 20% income group. **How will this affect our analysis?**

In [27]:
# Compute unique education level entries
philippine_data['educ'].nunique()

3

## Checking for errors

To check for data errors, we sometimes check for missing data and duplicated entries.

In [28]:
# Check for missing data
philippine_data.isna().sum()

economy                  0
economycode              0
regionwb                 0
pop_adult                0
wpid_random              0
                      ... 
remittances              0
pay_onlne                0
pay_onlne_mobintbuy    920
pay_cash                 0
pay_cash_mobintbuy     920
Length: 105, dtype: int64

In [29]:
# If we want, we can drop rows with NULLS in ANY columns
philippine_data.dropna()

Empty DataFrame
Columns: [economy, economycode, regionwb, pop_adult, wpid_random, wgt, female, age, educ, inc_q, emp_in, fin2, fin3, fin4, fin5, fin6, fin7, fin8, fin9, fin10, fin11a, fin11b, fin11c, fin11d, fin11e, fin11f, fin11g, fin11h, fin14a, fin14b, fin14c, fin15, fin16, fin17a, fin17b, fin19, fin20, fin21, fin22a, fin22b, fin22c, fin24, fin25, fin26, fin27a, fin27b, fin27c1, fin27c2, fin28, fin29a, fin29b, fin29c1, fin29c2, fin30, fin31a, fin31b, fin31c, fin32, fin33, fin34a, fin34b, fin34c1, fin34c2, fin35, fin36, fin37, fin38, fin39a, fin39b, fin39c1, fin39c2, fin40, fin41, fin42, fin43a, fin43b, fin43c1, fin43c2, fin44, fin45, fin46, fin47a, fin47b, fin47c1, fin47c2, fin47c3, fin47c4, fin47c5, mobileowner, fin48, account_fin, account_mob, account, saved, borrowed, receive_wages, receive_transfers, receive_pension, receive_agriculture, pay_utilities, ...]
Index: []

[0 rows x 105 columns]

In [30]:
# Or, we can only drop rows with NULLS in certain columns
philippine_data.dropna(subset=['pay_cash_mobintbuy'])

economy economycode                                     regionwb  \
114293  Philippines         PHL  East Asia & Pacific (excluding high income)   
114298  Philippines         PHL  East Asia & Pacific (excluding high income)   
114370  Philippines         PHL  East Asia & Pacific (excluding high income)   
114382  Philippines         PHL  East Asia & Pacific (excluding high income)   
114397  Philippines         PHL  East Asia & Pacific (excluding high income)   
...             ...         ...                                          ...   
115231  Philippines         PHL  East Asia & Pacific (excluding high income)   
115238  Philippines         PHL  East Asia & Pacific (excluding high income)   
115248  Philippines         PHL  East Asia & Pacific (excluding high income)   
115274  Philippines         PHL  East Asia & Pacific (excluding high income)   
115289  Philippines         PHL  East Asia & Pacific (excluding high income)   

         pop_adult  wpid_random       wgt  female   age  educ  inc_q  ...  \
114293  70293928.0    171987745  0.424916       2  24.0   3.0      5  ...   
114298  70293928.0    156608530  1.066926       1  18.0   2.0      2  ...   
114370  70293928.0    176096745  0.547007       1  54.0   3.0      5  ...   
114382  70293928.0    119360120  1.567463       1  17.0   2.0      2  ...   
114397  70293928.0    199330500  0.662789       2  16.0   1.0      2  ...   
...            ...          ...       ...     ...   ...   ...    ...  ...   
115231  70293928.0    163655673  0.331756       2  31.0   2.0      4  ...   
115238  70293928.0    118587065  0.534460       1  47.0   3.0      5  ...   
115248  70293928.0    199689458  0.356306       1  50.0   2.0      5  ...   
115274  70293928.0    120836654  1.496123       1  26.0   2.0      3  ...   
115289  70293928.0    134514859  2.614345       2  18.0   2.0      4  ...   

        receive_wages  receive_transfers  receive_pension  \
114293              1                  4                4   
114298              4                  4                4   
114370              4                  4                4   
114382              4                  4                4   
114397              4                  1                4   
...               ...                ...              ...   
115231              2                  1                4   
115238              4                  3                4   
115248              4                  3                4   
115274              1                  3                4   
115289              2                  4                4   

        receive_agriculture  pay_utilities  remittances  pay_onlne  \
114293                  4.0              4          1.0        0.0   
114298                  4.0              4          5.0        0.0   
114370                  2.0              2          2.0        0.0   
114382                  4.0              4          2.0        0.0   
114397                  4.0              4          5.0        0.0   
...                     ...            ...          ...        ...   
115231                  4.0              2          1.0        0.0   
115238                  4.0              1          5.0        0.0   
115248                  4.0              4          2.0        0.0   
115274                  2.0              4          2.0        0.0   
115289                  4.0              4          2.0        0.0   

        pay_onlne_mobintbuy  pay_cash  pay_cash_mobintbuy  
114293                  0.0       1.0                 1.0  
114298                  0.0       1.0                 1.0  
114370                  0.0       1.0                 1.0  
114382                  0.0       1.0                 1.0  
114397                  0.0       1.0                 1.0  
...                     ...       ...                 ...  
115231                  0.0       1.0                 1.0  
115238                  0.0       1.0                 1.0  
115248      

In [31]:
# Check for duplicated data in wpid_random (a unique identifier)
philippine_data.duplicated(subset='wpid_random', keep=False).sum()

0

**What does it mean?**

The subset checks for duplicated entries in wpid_random column. In this case, keep=False. Say, we have 4 duplicated entries. The keep=False parameter implies that we will mark ALL of these entries as duplicated. Thus, if we apply the .sum() method, it would result to 4.

Since there is no duplicated data in wpid_random column, we are sure that EACH respondent is unique!

In [32]:
# Save the data
philippine_data.to_csv(
    'philippine_data.csv',
    index=False
    )

To save the data, use .to_csv() or .to_excel() depending on what file format you need.

Here, index=False because I don't want to save the index as column in my .csv file.

## Sample exercises

In [39]:
# Select all female respondents in the Philippines
female_respondents = data[(data["female"] == 1) & (data["economy"] == "Philippines")]
female_respondents

economy economycode                                     regionwb  \
114297  Philippines         PHL  East Asia & Pacific (excluding high income)   
114298  Philippines         PHL  East Asia & Pacific (excluding high income)   
114303  Philippines         PHL  East Asia & Pacific (excluding high income)   
114306  Philippines         PHL  East Asia & Pacific (excluding high income)   
114307  Philippines         PHL  East Asia & Pacific (excluding high income)   
...             ...         ...                                          ...   
115282  Philippines         PHL  East Asia & Pacific (excluding high income)   
115287  Philippines         PHL  East Asia & Pacific (excluding high income)   
115288  Philippines         PHL  East Asia & Pacific (excluding high income)   
115290  Philippines         PHL  East Asia & Pacific (excluding high income)   
115291  Philippines         PHL  East Asia & Pacific (excluding high income)   

         pop_adult  wpid_random       wgt  female   age  educ  inc_q  ...  \
114297  70293928.0    187084257  1.197992       1  57.0   1.0      1  ...   
114298  70293928.0    156608530  1.066926       1  18.0   2.0      2  ...   
114303  70293928.0    132368001  0.916886       1  20.0   2.0      5  ...   
114306  70293928.0    126789618  2.441176       1  48.0   2.0      3  ...   
114307  70293928.0    130542330  0.261434       1  29.0   2.0      5  ...   
...            ...          ...       ...     ...   ...   ...    ...  ...   
115282  70293928.0    117412549  0.715865       1  27.0   2.0      4  ...   
115287  70293928.0    151988960  1.249522       1  27.0   2.0      1  ...   
115288  70293928.0    138300933  0.697020       1  68.0   1.0      1  ...   
115290  70293928.0    145841428  1.916246       1  20.0   2.0      4  ...   
115291  70293928.0    208872557  0.435442       1  27.0   1.0      1  ...   

        receive_wages  receive_transfers  receive_pension  \
114297              4                  4                4   
114298              4                  4                4   
114303              4                  3                4   
114306              4                  4                4   
114307              4                  4                4   
...               ...                ...              ...   
115282              2                  4                4   
115287              2                  4                4   
115288              4                  4                4   
115290              2                  4                4   
115291              2                  4                4   

        receive_agriculture  pay_utilities  remittances  pay_onlne  \
114297                  4.0              4          5.0        0.0   
114298                  4.0              4          5.0        0.0   
114303                  4.0              4          1.0        0.0   
114306                  2.0              2          5.0        0.0   
114307                  4.0              4          2.0        0.0   
...                     ...            ...          ...        ...   
115282                  2.0              2          3.0        0.0   
115287                  2.0              2          5.0        0.0   
115288                  4.0              4          5.0        0.0   
115290                  4.0              2          2.0        0.0   
115291                  2.0              4          2.0        0.0   

        pay_onlne_mobintbuy  pay_cash  pay_cash_mobintbuy  
114297                  NaN       0.0                 NaN  
114298                  0.0       1.0                 1.0  
114303                  NaN       0.0                 NaN  
114306                  NaN       0.0                 NaN  
114307                  NaN       0.0                 NaN  
...                     ...       ...                 ...  
115282                  NaN       0.0                 NaN  
115287                  NaN       0.0                 NaN  
115288      

In [41]:
# Select all female respondents with age < 30 
female_30 = data[(data["female"] == 1) & (data["age"] > 30)]
female_30

economy economycode                                    regionwb  \
5       Afghanistan         AFG                                  South Asia   
6       Afghanistan         AFG                                  South Asia   
9       Afghanistan         AFG                                  South Asia   
20      Afghanistan         AFG                                  South Asia   
24      Afghanistan         AFG                                  South Asia   
...             ...         ...                                         ...   
154916     Zimbabwe         ZWE  Sub-Saharan Africa (excluding high income)   
154919     Zimbabwe         ZWE  Sub-Saharan Africa (excluding high income)   
154920     Zimbabwe         ZWE  Sub-Saharan Africa (excluding high income)   
154921     Zimbabwe         ZWE  Sub-Saharan Africa (excluding high income)   
154922     Zimbabwe         ZWE  Sub-Saharan Africa (excluding high income)   

         pop_adult  wpid_random       wgt  female   age  educ  inc_q  ...  \
5       19456060.0    209050144  0.773167       1  53.0   2.0      1  ...   
6       19456060.0    167258084  0.481786       1  36.0   2.0      2  ...   
9       19456060.0    145691120  0.675545       1  75.0   1.0      5  ...   
20      19456060.0    156269367  0.360859       1  31.0   1.0      5  ...   
24      19456060.0    184272822  2.596681       1  42.0   1.0      4  ...   
...            ...          ...       ...     ...   ...   ...    ...  ...   
154916   9467437.0    188653559  0.811786       1  78.0   2.0      1  ...   
154919   9467437.0    121992603  0.576707       1  56.0   1.0      4  ...   
154920   9467437.0    137769624  1.481972       1  36.0   2.0      4  ...   
154921   9467437.0    181215643  0.259968       1  35.0   3.0      5  ...   
154922   9467437.0    133939668  0.791151       1  46.0   2.0      1  ...   

        receive_wages  receive_transfers  receive_pension  \
5                   2                  4                4   
6                   4                  2                4   
9                   4                  4                4   
20                  4                  4                4   
24                  2                  4                4   
...               ...                ...              ...   
154916              4                  1                1   
154919              4                  4                4   
154920              4                  4                4   
154921              1                  4                4   
154922              4                  3                4   

        receive_agriculture  pay_utilities  remittances  pay_onlne  \
5                       2.0              4          5.0        0.0   
6                       2.0              4          5.0        0.0   
9                       2.0              4          5.0        0.0   
20                      4.0              4          3.0        0.0   
24                      4.0              4          5.0        0.0   
...                     ...            ...          ...        ...   
154916                  4.0              4          5.0        0.0   
154919                  4.0              4          2.0        0.0   
154920                  4.0              4          1.0        0.0   
154921                  3.0              1          1.0        0.0   
154922                  4.0              4          3.0        0.0   

        pay_onlne_mobintbuy  pay_cash  pay_cash_mobintbuy  
5                       NaN       0.0                 NaN  
6                       NaN       0.0                 NaN  
9                       NaN       0.0                 NaN  
20                      NaN       0.0                 NaN  
24                      NaN       0.0                 NaN  
...                     ...       ...                 ...  
154916                  NaN       0.0                 NaN  
154919                  NaN       0.0                 NaN  
154920                  

In [45]:
# Select all female respondents that were able to complete secondary OR tertiary education 
female_education = data[(data["female"] == 1) & (data["educ"] == 2 | 3)]
female_education

economy economycode                                    regionwb  \
43      Afghanistan         AFG                                  South Asia   
86      Afghanistan         AFG                                  South Asia   
95      Afghanistan         AFG                                  South Asia   
116     Afghanistan         AFG                                  South Asia   
132     Afghanistan         AFG                                  South Asia   
...             ...         ...                                         ...   
154750     Zimbabwe         ZWE  Sub-Saharan Africa (excluding high income)   
154769     Zimbabwe         ZWE  Sub-Saharan Africa (excluding high income)   
154874     Zimbabwe         ZWE  Sub-Saharan Africa (excluding high income)   
154893     Zimbabwe         ZWE  Sub-Saharan Africa (excluding high income)   
154921     Zimbabwe         ZWE  Sub-Saharan Africa (excluding high income)   

         pop_adult  wpid_random       wgt  female   age  educ  inc_q  ...  \
43      19456060.0    199151425  0.648653       1  23.0   3.0      4  ...   
86      19456060.0    142048868  0.297843       1  30.0   3.0      4  ...   
95      19456060.0    203894824  0.267717       1  32.0   3.0      1  ...   
116     19456060.0    128521218  0.698141       1  31.0   3.0      5  ...   
132     19456060.0    146368347  1.936754       1  24.0   3.0      5  ...   
...            ...          ...       ...     ...   ...   ...    ...  ...   
154750   9467437.0    132507453  0.419116       1  27.0   3.0      5  ...   
154769   9467437.0    136856216  0.906158       1  37.0   3.0      5  ...   
154874   9467437.0    200329603  0.986287       1  21.0   3.0      5  ...   
154893   9467437.0    183043859  1.003786       1  29.0   3.0      2  ...   
154921   9467437.0    181215643  0.259968       1  35.0   3.0      5  ...   

        receive_wages  receive_transfers  receive_pension  \
43                  4                  4                4   
86                  1                  4                4   
95                  2                  4                4   
116                 1                  4                4   
132                 4                  4                4   
...               ...                ...              ...   
154750              4                  4                4   
154769              1                  4                4   
154874              1                  4                4   
154893              1                  4                4   
154921              1                  4                4   

        receive_agriculture  pay_utilities  remittances  pay_onlne  \
43                      4.0              4          3.0        0.0   
86                      1.0              1          1.0        0.0   
95                      4.0              4          5.0        0.0   
116                     4.0              4          5.0        0.0   
132                     1.0              2          5.0        0.0   
...                     ...            ...          ...        ...   
154750                  4.0              1          1.0        0.0   
154769                  4.0              1          1.0        0.0   
154874                  4.0              2          1.0        0.0   
154893                  4.0              1          2.0        0.0   
154921                  3.0              1          1.0        0.0   

        pay_onlne_mobintbuy  pay_cash  pay_cash_mobintbuy  
43                      NaN       0.0                 NaN  
86                      NaN       0.0                 NaN  
95                      NaN       0.0                 NaN  
116                     NaN       0.0                 NaN  
132                     NaN       0.0                 NaN  
...                     ...       ...                 ...  
154750                  NaN       0.0                 NaN  
154769                  NaN       0.0                 NaN  
154874                  